atom/browser/api/atom_api_net_log.cc

```diff
@@ -21,103 +22,145 @@
+  return base::CreateSequencedTaskRunnerWithTraits(
+      {base::MayBlock(), base::TaskPriority::USER_VISIBLE,
+       base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN});
+}
+
+base::File OpenFileForWriting(base::FilePath path) {
+  return base::File(path,
+                    base::File::FLAG_CREATE_ALWAYS | base::File::FLAG_WRITE);
+}
+
+void ResolvePromiseWithNetError(util::Promise promise, int32_t error) {
+  if (error == net::OK) {
+    promise.Resolve();
+  } else {
+    promise.RejectWithErrorMessage(net::ErrorToString(error));
+  }
+}
+
+}  // namespace
+
 namespace api {
 
 NetLog::NetLog(v8::Isolate* isolate, AtomBrowserContext* browser_context)
-    : browser_context_(browser_context) {
+    : browser_context_(browser_context), weak_ptr_factory_(this) {
   Init(isolate);
-
-  net_log_writer_ = g_browser_process->system_network_context_manager()
-                        ->GetNetExportFileWriter();
-  net_log_writer_->AddObserver(this);
+  file_task_runner_ = CreateFileTaskRunner();
 }
 
-NetLog::~NetLog() {
-  net_log_writer_->RemoveObserver(this);
-}
+NetLog::~NetLog() = default;
 
-void NetLog::StartLogging(mate::Arguments* args) {
+v8::Local<v8::Promise> NetLog::StartLogging(mate::Arguments* args) {
   base::FilePath log_path;
   if (!args->GetNext(&log_path) || log_path.empty()) {
     args->ThrowError("The first parameter must be a valid string");
-    return;
+    return v8::Local<v8::Promise>();
+  }
+
+  if (net_log_exporter_) {
+    args->ThrowError("There is already a net log running");
+    return v8::Local<v8::Promise>();
   }
 
+  util::Promise promise(isolate());
+  v8::Local<v8::Promise> handle = promise.GetHandle();
+
+  auto command_line_string =
+      base::CommandLine::ForCurrentProcess()->GetCommandLineString();
+  auto channel_string = std::string("Electron " ATOM_VERSION);
+  base::Value custom_constants = base::Value::FromUniquePtrValue(
+      net_log::ChromeNetLog::GetPlatformConstants(command_line_string,
+                                                  channel_string));
+
   auto* network_context =
       content::BrowserContext::GetDefaultStoragePartition(browser_context_)
           ->GetNetworkContext();
 
+  network_context->CreateNetLogExporter(mojo::MakeRequest(&net_log_exporter_));
+  net_log_exporter_.set_connection_error_handler(
+      base::BindOnce(&NetLog::OnConnectionError, base::Unretained(this)));
+
   // TODO(deepak1556): Provide more flexibility to this module
   // by allowing customizations on the capturing options.
-  net_log_writer_->StartNetLog(
-      log_path, net::NetLogCaptureMode::Default(),
-      net_log::NetExportFileWriter::kNoLimit /* file size limit */,
-      base::CommandLine::ForCurrentProcess()->GetCommandLineString(),
-      std::string(), network_context);
-}
+  auto capture_mode = network::mojom::NetLogCaptureMode::DEFAULT;
+  auto max_file_size = network::mojom::NetLogExporter::kUnlimitedFileSize;
```

Comment:
Can we inline these at the call site, I don't see the variables being much useful. Any reason to have these declared separately ?

Reply:
My reason for breaking these out is that I plan to make them configurable in a follow-up PR.